elements/lisk-chain/src/chain.ts

```diff
@@ -376,78 +385,58 @@ export class Chain {
-		});
-		const failedAllowedResponses = allowedResponses.filter(
-			res => res.status === TransactionStatus.FAIL,
+	public async getValidators(): Promise<Validator[]> {
+		const validatorsBuffer = await this.dataAccess.getConsensusState(
+			CONSENSUS_STATE_VALIDATORS_KEY,
 		);
-		const validTransactions = transactions.filter(tx =>
-			allowedResponses.find(res => res.status === TransactionStatus.OK && res.id.equals(tx.id)),
+		if (!validatorsBuffer) {
+			return [];
+		}
+		const { validators } = codec.decode<{ validators: Validator[] }>(
+			validatorsSchema,
+			validatorsBuffer,
 		);
-		const validationResponses = validateTransactions(validTransactions);
-		return [...failedAllowedResponses, ...validationResponses];
-	}
 
-	public async applyTransactions(transactions: BaseTransaction[]): Promise<TransactionResponse[]> {
-		const stateStore = await this.newStateStore();
-		const allowedResponses = checkAllowedTransactions(transactions, () => {
-			const { version, height, timestamp } = this._lastBlock.header;
-			return {
-				blockVersion: version,
-				blockHeight: height,
-				blockTimestamp: timestamp,
-			};
-		});
-		const failedAllowedResponses = allowedResponses.filter(
-			res => res.status === TransactionStatus.FAIL,
-		);
-		const validTransactions = transactions.filter(tx =>
-			allowedResponses.find(res => res.status === TransactionStatus.OK && res.id.equals(tx.id)),
-		);
-		const applyResponses = await applyTransactions(validTransactions, stateStore);
-		return [...failedAllowedResponses, ...applyResponses];
+		return validators;
 	}
 
 	// eslint-disable-next-line class-methods-use-this
-	public async applyTransactionsWithStateStore(
-		transactions: BaseTransaction[],
+	public async setValidators(
+		validators: { address: Buffer; isConsensusParticipant: boolean }[],
 		stateStore: StateStore,
-	): Promise<ReadonlyArray<TransactionResponse>> {
-		return applyTransactions(transactions, stateStore);
+	): Promise<void> {
+		const validatorsBuffer = await stateStore.consensus.get(CONSENSUS_STATE_VALIDATORS_KEY);
+		if (!validatorsBuffer) {
+			const initialValidators = validators.map(v => ({ ...v, minActiveHeight: 0 }));
+			stateStore.consensus.set(
+				CONSENSUS_STATE_VALIDATORS_KEY,
+				codec.encode(validatorsSchema, { validators: initialValidators }),
+			);
+			return;
+		}
+		const { validators: previousValidators } = codec.decode<{ validators: Validator[] }>(
+			validatorsSchema,
+			validatorsBuffer,
+		);
+		const [lastBlock] = stateStore.chain.lastBlockHeaders;
+		const nextValidatorSet = [];
+		for (const nextValidator of validators) {
+			const previousInfo = previousValidators.find(pv => pv.address.equals(nextValidator.address));
+			nextValidatorSet.push({
+				...nextValidator,
+				minActiveHeight:
+					// currently processed block is still in process, so the next block height is +2
+					previousInfo !== undefined ? previousInfo.minActiveHeight : lastBlock.height + 2,
```

Comment:
dint quite get this +2 logic, could you detail it out a bit more?

Reply:
`currently processed block is still in process, so the next block height is +2`
meaning when this function is called `| last block | currently processed block | next block|` and what we want is the height of the next block